File: app/src/main/java/com/ternaryop/lazyimageloader/ImageLoader.java
package com.ternaryop.lazyimageloader;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.util.Collections;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.view.MenuItem;
import android.widget.ImageView;

import com.ternaryop.utils.ImageUtils;
import com.ternaryop.utils.URLUtils;

public class ImageLoader {
    public final static String IMAGE_PREFIX_DIRECTORY = "images" + File.separator;
    
    private final MemoryCache memoryCache = new MemoryCache();
    private final FileCache fileCache;
    private final Map<ImageView, String> imageViews = Collections.synchronizedMap(new WeakHashMap<ImageView, String>());
    private final ExecutorService executorService;
    private final Handler handler = new Handler(); // handler to display images in UI thread
    private final int stubResId;
    private final Context context;

    public ImageLoader(Context context, String prefix, int stubResId) {
        this.context = context;
        this.fileCache = new FileCache(context, IMAGE_PREFIX_DIRECTORY + prefix);
        this.stubResId = stubResId;
        executorService = Executors.newFixedThreadPool(5);
    }

    public void displayImage(String url, ImageView imageView) {
        displayImage(url, imageView, false);
    }

    public void displayImage(String url, ImageView imageView, boolean scaleForDefaultDisplay) {
        imageViews.put(imageView, url);
        Bitmap bitmap = memoryCache.get(url);
        if (bitmap != null) {
            if (scaleForDefaultDisplay) {
                bitmap = ImageUtils.scaleBitmapForDefaultDisplay(context, bitmap);
            }
            imageView.setImageBitmap(bitmap);
        } else {
            queuePhoto(url, imageView, scaleForDefaultDisplay);
            imageView.setImageResource(stubResId);
        }
    }
    
    public void displayDrawable(final String url, final ImageLoaderCallback callback) {
        displayDrawable(url, callback, true);
    }
    
    public void displayDrawable(final String url, final ImageLoaderCallback callback, final boolean scaleForDefaultDisplay) {
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                final Drawable icon = drawableFromUrl(url, scaleForDefaultDisplay);
                if (icon != null) {
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            callback.display(icon);
                        }
                    });
                }
            }
        });
    }
    
    public void displayIcon(final String url, final MenuItem menuItem) {
        displayDrawable(url, new ImageLoaderCallback() {
            @Override
            public void display(Drawable drawable) {
                menuItem.setIcon(drawable);
            }
        });
    }
        
    private void queuePhoto(String url, ImageView imageView, boolean scaleForDefaultDisplay)
    {
        PhotoToLoad p=new PhotoToLoad(url, imageView);
        executorService.submit(new PhotosLoader(p, scaleForDefaultDisplay));
    }
    
    private Bitmap getBitmap(String url)  {
        File f = fileCache.getFile(url);
        
        //from SD cache
        Bitmap b = decodeFile(f);
        if (b != null) {
            return b;
        }

        //from web
        try {
            HttpURLConnection conn = URLUtils.openConnectionFollowingDifferentProtocols(url);
            conn.setConnectTimeout(30000);
            conn.setReadTimeout(30000);
            conn.connect();
            InputStream is = conn.getInputStream();
            OutputStream os = new FileOutputStream(f);
            Utils.CopyStream(is, os);
            os.close();
            conn.disconnect();
            return decodeFile(f);
        } catch (Throwable ex){
           ex.printStackTrace();
           if(ex instanceof OutOfMemoryError)
               memoryCache.clear();
           return null;
        }
    }

    //decodes image and scales it to reduce memory consumption
    private Bitmap decodeFile(File f){
        try {
            //decode image size
            BitmapFactory.Options o = new BitmapFactory.Options();
            o.inJustDecodeBounds = true;
            FileInputStream stream1=new FileInputStream(f);
            BitmapFactory.decodeStream(stream1,null,o);
            stream1.close();
            
            //Find the correct scale value. It should be the power of 2.
            final int REQUIRED_SIZE=70;
            int width_tmp=o.outWidth, height_tmp=o.outHeight;
            int scale=1;
            while(true){
                if(width_tmp/2<REQUIRED_SIZE || height_tmp/2<REQUIRED_SIZE)
                    break;
                width_tmp/=2;
                height_tmp/=2;
                scale*=2;
            }
            
            //decode with inSampleSize
            BitmapFactory.Options o2 = new BitmapFactory.Options();
            o2.inSampleSize=scale;
            FileInputStream stream2=new FileInputStream(f);
            Bitmap bitmap=BitmapFactory.decodeStream(stream2, null, o2);
            stream2.close();
            return bitmap;
        } catch (FileNotFoundException ignored) {
        } 
        catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
    
    //Task for the queue
    private class PhotoToLoad
    {
        public final String url;
        public final ImageView imageView;
        public PhotoToLoad(String u, ImageView i){
            url=u; 
            imageView=i;
        }
    }
    
    class PhotosLoader implements Runnable {
        final PhotoToLoad photoToLoad;
        private final boolean scaleForDefaultDisplay;
        PhotosLoader(PhotoToLoad photoToLoad, boolean scaleForDefaultDisplay) {
            this.photoToLoad = photoToLoad;
            this.scaleForDefaultDisplay = scaleForDefaultDisplay;
        }
        
        @Override
        public void run() {
            try {
                if (imageViewReused(photoToLoad)) {
                    return;
                }
                Bitmap bmp = getBitmap(photoToLoad.url);
                memoryCache.put(photoToLoad.url, bmp);
                if (imageViewReused(photoToLoad)) {
                    return;
                }
                BitmapDisplayer bd = new BitmapDisplayer(bmp, photoToLoad, scaleForDefaultDisplay);
                handler.post(bd);
            } catch (Throwable th) {
                th.printStackTrace();
            }
        }
    }
    
    boolean imageViewReused(PhotoToLoad photoToLoad){
        String tag=imageViews.get(photoToLoad.imageView);
        return tag == null || !tag.equals(photoToLoad.url);
    }
    
    //Used to display bitmap in the UI thread
    class BitmapDisplayer implements Runnable {
        Bitmap bitmap;
        final PhotoToLoad photoToLoad;
        private final boolean scaleForDefaultDisplay;

        public BitmapDisplayer(Bitmap b, PhotoToLoad p, boolean scaleForDefaultDisplay) {
            bitmap = b;
            photoToLoad = p;
            this.scaleForDefaultDisplay = scaleForDefaultDisplay;
        }

        public void run()
        {
            if (imageViewReused(photoToLoad)) {
                return;
            }
            if (bitmap != null) {
                if (scaleForDefaultDisplay) {
                    bitmap = ImageUtils.scaleBitmapForDefaultDisplay(context, bitmap);
                }
                photoToLoad.imageView.setImageBitmap(bitmap);
            } else {
                photoToLoad.imageView.setImageResource(stubResId);
            }
        }
    }

    public void clearCache() {
        memoryCache.clear();
        fileCache.clear();
    }

    public static void clearImageCache(Context context) {
        FileCache.clearCache(context, IMAGE_PREFIX_DIRECTORY);
    }
    
    protected Drawable drawableFromUrl(String url, boolean scaleForDefaultDisplay) {
        HttpURLConnection conn = null;
        OutputStream os = null;
        try {
            File f = fileCache.getFile(url);
            Bitmap bitmap = Utils.decodeFile(f);

            if (bitmap == null) {
                conn = URLUtils.openConnectionFollowingDifferentProtocols(url);
                conn.setConnectTimeout(30000);
                conn.setReadTimeout(30000);
                conn.connect();
                os = new FileOutputStream(f);
                Utils.CopyStream(conn.getInputStream(), os);
                
                bitmap = Utils.decodeFile(f);
            }
            if (scaleForDefaultDisplay) {
                bitmap = ImageUtils.scaleBitmapForDefaultDisplay(context, bitmap);
            }
            return new BitmapDrawable(null, bitmap);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try { if (os != null) os.close(); } catch (Exception ignored) {}
            try { if (conn != null) conn.disconnect(); } catch (Exception ignored) {}
        }
        return null;
    }

    public interface ImageLoaderCallback {
        public void display(Drawable drawable);
    }
}

File: app/src/main/java/com/ternaryop/lazyimageloader/ImageLoader.kt
package com.ternaryop.lazyimageloader

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Handler
import android.view.MenuItem
import android.widget.ImageView
import com.ternaryop.utils.bitmap.scaleForDefaultDisplay
import com.ternaryop.utils.network.openConnectionFollowingDifferentProtocols
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.Collections
import java.util.WeakHashMap
import java.util.concurrent.Executors

class ImageLoader(private val context: Context, prefix: String, private val stubResId: Int) {
    private val memoryCache = MemoryCache()
    private val fileCache = FileCache(context, IMAGE_PREFIX_DIRECTORY + prefix)
    private val imageViews = Collections.synchronizedMap(WeakHashMap<ImageView, String>())
    private val executorService = Executors.newFixedThreadPool(5)
    private val handler = Handler() // handler to display images in UI thread

    fun displayImage(url: String, imageView: ImageView, scaleForDefaultDisplay: Boolean = false) {
        imageViews[imageView] = url
        var bitmap = memoryCache[url]
        if (bitmap != null) {
            if (scaleForDefaultDisplay) {
                bitmap = bitmap.scaleForDefaultDisplay(context)
            }
            imageView.setImageBitmap(bitmap)
        } else {
            queuePhoto(url, imageView, scaleForDefaultDisplay)
            imageView.setImageResource(stubResId)
        }
    }

    fun displayDrawable(url: String, scaleForDefaultDisplay: Boolean, callback: ImageLoaderCallback) {
        executorService.submit {
            drawableFromUrl(url, scaleForDefaultDisplay)?.let { icon -> handler.post { callback.display(icon) } }
        }
    }

    fun displayIcon(url: String, menuItem: MenuItem) {
        displayDrawable(url, true, object : ImageLoaderCallback {
            override fun display(drawable: Drawable) {
                menuItem.icon = drawable
            }
        })
    }

    private fun queuePhoto(url: String, imageView: ImageView, scaleForDefaultDisplay: Boolean) {
        val p = PhotoToLoad(url, imageView)
        executorService.submit(PhotosLoader(p, scaleForDefaultDisplay))
    }

    private fun getBitmap(url: String): Bitmap? {
        val f = fileCache.getFile(url)
        //from SD cache
        val b = decodeFile(f)
        if (b != null) {
            return b
        }
        //from web
        try {
            val conn = URL(url).openConnectionFollowingDifferentProtocols()
            conn.connectTimeout = 30000
            conn.readTimeout = 30000
            conn.connect()
            val stream = conn.inputStream
            val os = FileOutputStream(f)
            Utils.copyStream(stream, os)
            os.close()
            conn.disconnect()
            return decodeFile(f)
        } catch (ex: Throwable) {
            ex.printStackTrace()
            if (ex is OutOfMemoryError)
                memoryCache.clear()
            return null
        }
    }

    //decodes image and scales it to reduce memory consumption
    private fun decodeFile(f: File): Bitmap? {
        try {
            //decode image size
            val o = BitmapFactory.Options()
            o.inJustDecodeBounds = true
            val stream1 = FileInputStream(f)
            BitmapFactory.decodeStream(stream1, null, o)
            stream1.close()
            //Find the correct scale value. It should be the power of 2.
            val requiredSize = 70
            var tmpWidth = o.outWidth
            var tmpHeight = o.outHeight
            var scale = 1
            while (true) {
                if (tmpWidth / 2 < requiredSize || tmpHeight / 2 < requiredSize)
                    break
                tmpWidth /= 2
                tmpHeight /= 2
                scale *= 2
            }
            //decode with inSampleSize
            val o2 = BitmapFactory.Options()
            o2.inSampleSize = scale
            val stream2 = FileInputStream(f)
            val bitmap = BitmapFactory.decodeStream(stream2, null, o2)
            stream2.close()
            return bitmap
        } catch (ignored: FileNotFoundException) {
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    //Task for the queue
    private inner class PhotoToLoad(val url: String, val imageView: ImageView)

    private inner class PhotosLoader(val photoToLoad: PhotoToLoad, private val scaleForDefaultDisplay: Boolean) : Runnable {
        override fun run() {
            try {
                if (imageViewReused(photoToLoad)) {
                    return
                }
                val bmp = getBitmap(photoToLoad.url)
                memoryCache.put(photoToLoad.url, bmp!!)
                if (imageViewReused(photoToLoad)) {
                    return
                }
                val bd = BitmapDisplayer(bmp, photoToLoad, scaleForDefaultDisplay)
                handler.post(bd)
            } catch (th: Throwable) {
                th.printStackTrace()
            }
        }
    }

    private fun imageViewReused(photoToLoad: PhotoToLoad): Boolean {
        val tag = imageViews[photoToLoad.imageView]
        return tag == null || tag != photoToLoad.url
    }

    //Used to display bitmap in the UI thread
    private inner class BitmapDisplayer(val bitmap: Bitmap?, val photoToLoad: PhotoToLoad, private val scaleForDefaultDisplay: Boolean) : Runnable {
        override fun run() {
            if (imageViewReused(photoToLoad)) {
                return
            }
            if (bitmap != null) {
                photoToLoad.imageView.setImageBitmap(if (scaleForDefaultDisplay)
                    bitmap.scaleForDefaultDisplay(context) else bitmap)
            } else {
                photoToLoad.imageView.setImageResource(stubResId)
            }
        }
    }

    fun clearCache() {
        memoryCache.clear()
        fileCache.clear()
    }

    private fun drawableFromUrl(url: String, scaleForDefaultDisplay: Boolean): Drawable? {
        var conn: HttpURLConnection? = null
        var os: OutputStream? = null
        try {
            val f = fileCache.getFile(url)
            var bitmap = Utils.decodeFile(f)

            if (bitmap == null) {
                conn = URL(url).openConnectionFollowingDifferentProtocols()
                conn.connectTimeout = 30000
                conn.readTimeout = 30000
                conn.connect()
                os = FileOutputStream(f)
                Utils.copyStream(conn.inputStream, os)

                bitmap = Utils.decodeFile(f)
            }
            if (scaleForDefaultDisplay) {
                bitmap = bitmap?.scaleForDefaultDisplay(context)
            }
            return BitmapDrawable(null, bitmap)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                if (os != null) os.close()
            } catch (ignored: Exception) {
            }

            try {
                if (conn != null) conn.disconnect()
            } catch (ignored: Exception) {
            }
        }
        return null
    }

    interface ImageLoaderCallback {
        fun display(drawable: Drawable)
    }

    companion object {
        val IMAGE_PREFIX_DIRECTORY = "images" + File.separator

        fun clearImageCache(context: Context) {
            FileCache.clearCache(context, IMAGE_PREFIX_DIRECTORY)
        }
    }
}
